### ISSUE TITLE
[C_API] The same LGBM Model returns different predictions on different platforms

### ISSUE THREAD

**Issue by JavahirAbbasova** [microsoft/LightGBM#5843]
## Description
We are using LGBMRegressor in generation of various customer specific models. The models are trained and hosted by MLFlow Python library. 

The lifecycle of the models is as follows:
- The models are trained by execution of python scripts locally.
- The MLFlow artifacts including the model as .pkl file are deployed to the cloud.
- The models are hosted in the cloud and respond to prediction queries. 

We have been observing differences between predictions coming from the same LGBM model hosted in the cloud and on our local machines. For example, for the same data the cloud returns [78.572706, 77.618487], while we get [68.638522, 64.589452] locally.

## Reproducible example
Unfortunately, no reproducible example can be provided due to sensitive customer data. 

## Environment info

LightGBM version: 3.3.3

Command(s) to install LightGBM

LightGBM has been installed with pip in a conda environment.

## Additional Comments
Cloud: 64-Bit-Linux
Local: 64-Bit-Windows

We could verify that the different OS is not the cause of the problem, as the model hosted in Docker returns the same predictions we get locally. We could also confirm through debugging that the differences don't occur in the data or model in previous steps. In fact, we could identify only one difference between the two versions of the same model: For the model hosted in the cloud the variable `out_num_preds` (the function `inner_predict` in basic.py) has the type c_long, while in the model hosted locally the type is `c_longlong`. We are assuming that the platform sensitive `cytpes` may cause the problem we are facing.

**Comment by jmoralez** [comment]
Hi @JavahirAbbasova, thanks for using LightGBM. Unfortunately, without a reproducible example this is pretty much impossible for us to debug. For starters we can't even tell if there is indeed a problem in LightGBM itself or in a different part of your workflow. It could be due to different python versions (e.g. differences related to pickle), versions of libraries, etc.

If you're able to produce a minimal reproducible example (doesn't have to include any of your customer data, just an example where you observe the same behavior) we can provide further assistance, otherwise it's very difficult for us to help.

**Comment by JavahirAbbasova** [comment]
Hi @jmoralez. Thank you for your reply. We could anonymize the customer data as well as reduce the code to the minimum to enable you the debugging. You would only need to create a conda environment (python=3.10) and install the modules listed in requirements.txt attached. Then unpack the LGBM_ISSUE.zip and run the test.py in your newly created conda environment. As described above, this code prints [68.638522, 64.589452] on our local machines and [78.572706, 77.618487] in the cloud. Feel free to contact me, if you need further details.

[requirements.txt](https://github.com/microsoft/LightGBM/files/11344529/requirements.txt)
[LGBM_ISSUE.zip](https://github.com/microsoft/LightGBM/files/11344531/LGBM_ISSUE.zip)

**Comment by jmoralez** [comment]
Hi @JavahirAbbasova, I ran this tonight on my laptop with ubuntu and I get [68.638522, 64.589452], so it doesn't seem to be related to ctypes. Are you sure you have the same versions of libraries on your cloud deployment?

**Comment by JavahirAbbasova** [comment]
Yes @jmoralez , the container image for hosting the model is built with a Dockerfile that installs the libraries before starting the application. Afterwards, the same image is delivered to the cloud and started there as a container. No additional libraries are installed or the existing ones updated in the cloud environment. Therefore, the library versions are exact the same in the docker container running locally and in the cloud. However, the container running locally returns [68.638522, 64.589452] while the container running in the cloud returns [78.572706, 77.618487].

**Comment by jmoralez** [comment]
Another thing that comes to mind is a difference in the CPU architectures. Maybe you're building the image in x86 and deploying it on arm64, like EC2 Graviton or similar.

**Comment by JavahirAbbasova** [comment]
Both build and deployment systems are x64.
